keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1342 @@
+  adopting upstream.
+
+[kube-nftlb]: https://github.com/zevenet/kube-nftlb
+[nftlb]: https://github.com/zevenet/nftlb
+[nfproxy]: https://github.com/sbezverk/nfproxy
+[kpng's nft backend]: https://github.com/kubernetes-sigs/kpng/tree/master/backends/nft
+
+### Risks and Mitigations
+
+The primary risk of the proposal is feature regressions, which will be
+addressed by testing, and by a slow, optional, rollout of the new proxy
+mode.
+
+The `nftables` mode should not pose any new security issues relative
+to the `iptables` mode.
+
+## Design Details
+
+### High level
+
+At a high level, the new mode should have the same architecture as the
+existing modes; it will use the service/endpoint-tracking code in
+`k8s.io/kubernetes/pkg/proxy` (or its eventual replacement from kpng)
+to watch for changes, and update rules in the kernel accordingly.
+
+### Low level
+
+Some details will be figured out as we implement it. We may start with
+an implementation that is architecturally closer to the `iptables`
+mode, and then rewrite it to take advantage of additional nftables
+features over time.
+
+#### Tables
+
+Unlike iptables, nftables does not have any reserved/default tables or
+chains (eg, `nat`, `PREROUTING`). Users are expected to create their
+own tables and chains for their own purposes. An nftables table can
+only contain rules for a single "family" (`ip` (v4), `ip6`, `inet`
+(both IPv4 and IPv6), `arp`, `bridge`, or `netdev`), but unlike in
+iptables, you can have both "filter"-type chains and "NAT"-type chains
+in the same table.
+
+So, we will create a single `kube_proxy` table in the `ip` family, and
+another in the `ip6` family. All of our chains, sets, maps, etc, will
+go into those tables. Other system components (eg, firewalld) should
+ignore our table, so we should not need to worry about watching for
+other people deleting our rules like we have to in the `iptables`
+backend.
+
+(In theory, instead of creating one table each in the `ip` and `ip6`
+families, we could create a single table in the `inet` family and put
+both IPv4 and IPv6 chains/rules there. However, this wouldn't really
+result in much simplification, because we would still need separate
+sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
+data type that can store/match either an IPv4 address or an IPv6
+address.) Furthermore, because of how Kubernetes Services evolved in
+parallel with the existing kube-proxy implementation, we have ended up
+with a dual-stack Service semantics that is most easily implemented by
+handling IPv4 and IPv6 completely separately anyway.)
+
+#### Communicating with the kernel nftables subsystem
+
+At least initially, we will use the `nft` command-line tool to read
+and write rules, much like how we use command-line tools in the
+`iptables` and `ipvs` backends. However, the `nft` tool is mostly just
+a thin wrapper around `libnftables`, and it would be possible to use
```

Comment:
we need to update the image for these new dependencies or create a new one
https://github.com/kubernetes/release/tree/master/images/build/distroless-iptables
```
 ldd /usr/sbin/nft
        linux-vdso.so.1 (0x00007ffc8cbcb000)
        libnftables.so.1 => /lib/x86_64-linux-gnu/libnftables.so.1 (0x00007feeb52d0000)
        libedit.so.2 => /lib/x86_64-linux-gnu/libedit.so.2 (0x00007feeb5296000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007feeb506e000)
        libmnl.so.0 => /lib/x86_64-linux-gnu/libmnl.so.0 (0x00007feeb5066000)
        libnftnl.so.11 => /lib/x86_64-linux-gnu/libnftnl.so.11 (0x00007feeb5034000)
        libxtables.so.12 => /lib/x86_64-linux-gnu/libxtables.so.12 (0x00007feeb5021000)
        libjansson.so.4 => /lib/x86_64-linux-gnu/libjansson.so.4 (0x00007feeb500f000)
        libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007feeb4f8d000)
        libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007feeb4f5b000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007feeb4f43000)
        /lib64/ld-linux-x86-64.so.2 (0x00007feeb53b5000)
        libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007feeb4f36000)
```

Reply:
(note that `iptables-nft` already depends on half of those things too... it looks like nft should add about 1M to the image)